### PR TITLE
Make Wikitext importer more robust

### LINKED ISSUE #1448: Error on Wikitable import using "Extract references in additional columns" option
Trying to import the following Wikitable:
```
{| class="wikitable" style="width:80%; text-align:center;"
|-
! President
! #
! width=70|Portrait
! width=25%|Name
! colspan="2" width=35%|Term of office
! Political Party
|-
| rowspan="2" align="center"|[[Carlos Menem]]
! 1
| [[File:Coat of arms of Argentina.svg|70px]]
| '''[[Eduardo Bauzá]]'''
| 8 July 1995
| 5 June 1996
| [[Justicialist Party]]
|-
! 2
| [[File:Coat of arms of Argentina.svg|70px]]
| '''[[Jorge Alberto Rodríguez]]'''
| 5 June 1996
| 10 December 1999 
| [[Justicialist Party]]
|-
| rowspan="2" align="center"|[[Fernando de la Rúa]]
! 3
| [[File:Rodolfoterragno.jpg|100px|border]]
| '''[[Rodolfo Terragno]]'''
| 10 December 1999
| 6 October 2000
| [[Radical Civic Union]]
|-
! 4
| [[File:Coat of arms of Argentina.svg|70px]]
| '''[[Chrystian Colombo]]'''
| 6 October 2000
| 20 December 2001
| [[Radical Civic Union]]
|-
| align="center"|[[Ramón Puerta]]
! 5
| [[File:Humberto Schiavoni.jpg|100px|border]]
| '''[[Humberto Schiavoni]]'''
| 20 December 2001
| 23 December 2001
| [[Justicialist Party]]
|-
| align="center"|[[Adolfo Rodríguez Saá]]
! 6
| [[File:Lusquiños.jpg|100px|border]]
| '''[[Luis Lusquiños]]'''
| 23 December 2001
| 30 December 2001
| [[Justicialist Party]]
|-
| align="center"|[[Eduardo Camaño]]
! 7
| [[File:Antonio Cafiero en 1975.jpg|100px|border]]
| '''[[Antonio Cafiero]]'''
| 30 December 2001
| 2 January 2002
| [[Justicialist Party]]
|-
| rowspan="2" align="center"|[[Eduardo Duhalde]]
! 8
| [[File:Jcapitanich.jpg|100px|border]]
| '''[[Jorge Capitanich]]'''
| 2 January 2002
| 3 May 2002
| [[Justicialist Party]]
|-
! 9
| [[File:Alfredo Atanasof.jpg|100px|border]]
| '''[[Alfredo Atanasof]]'''
| 3 May 2002
| 25 May 2003
| [[Justicialist Party]]
|-
| align="center"|[[Néstor Kirchner]]
! rowspan="2" | 10
| rowspan="2" | [[File:AFernandez.jpg|100px|border]]
| rowspan="2" | '''[[Alberto Ángel Fernández]]'''
| rowspan="2" | 25 May 2003
| rowspan="2" | 23 July 2008
| rowspan="2" rowspan="2" | [[Justicialist Party]]
|-
| rowspan="6" align="center"|[[Cristina Fernández de Kirchner]]
|-
! 11
| [[File:Sergio Massa (1).jpg|100px|border]]
| '''[[Sergio Massa]]'''
| 23 July 2008
| 7 July 2009
| [[Justicialist Party]]
|-
! 12
| [[File:Aníbal Fernández (2011).jpg|100px|border]]
| '''[[Aníbal Fernández]]'''
| 7 July 2009
| 10 December 2011
| [[Justicialist Party]]
|-
! 13
| [[File:JuanManuelAbalMedina.jpg|100px|border]]
| '''[[Juan Manuel Abal Medina]]'''
| 10 December 2011
| 20 December 2013
| [[Justicialist Party]]
|-
! 14
| [[File:Jcapitanich.jpg|100px|border]]
| '''[[Jorge Capitanich]]'''
| 20 December 2013
| 26 February 2015
| [[Justicialist Party]]
|-
! 15
| [[File:Aníbal Fernández (2011).jpg|100px|border]]
| '''[[Aníbal Fernández]]'''
| 26 February 2015
| 10 December 2015
| [[Justicialist Party]]
|-
| align="center"| [[Mauricio Macri]]
! 16
| [[File:Marcos Peña 2016 (cropped).jpg|100px|border]]
| '''[[Marcos Peña]]'''
| 10 December 2015
| ''present''
| [[Republican Proposal]]
|}
```

The in the UI it just hangs with the message "Updating Preview"
In the console the following error:

```
15:35:06.832 [                   refine] POST /command/core/importing-controller (11356ms)
15:35:09.191 [          org.mortbay.log] /command/core/importing-controller (2359ms)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 6
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at com.google.refine.importers.WikitextImporter$WikiTableDataReader.getNextRowOfCells(WikitextImporter.java:589)
	at com.google.refine.importers.TabularImportingParserBase.readTable(TabularImportingParserBase.java:121)
	at com.google.refine.importers.WikitextImporter.parseOneFile(WikitextImporter.java:716)
	at com.google.refine.importers.ImportingParserBase.parseOneFile(ImportingParserBase.java:128)
	at com.google.refine.importers.ImportingParserBase.parse(ImportingParserBase.java:89)
	at com.google.refine.importing.ImportingUtilities.previewParse(ImportingUtilities.java:956)
	at com.google.refine.importing.DefaultImportingController.doUpdateFormatAndOptions(DefaultImportingController.java:185)
	at com.google.refine.importing.DefaultImportingController.doPost(DefaultImportingController.java:93)
	at com.google.refine.commands.importing.ImportingControllerCommand.doPost(ImportingControllerCommand.java:62)
	at com.google.refine.RefineServlet.service(RefineServlet.java:177)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at org.mortbay.servlet.UserAgentFilter.doFilter(UserAgentFilter.java:81)
	at org.mortbay.servlet.GzipFilter.doFilter(GzipFilter.java:132)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:938)
	at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:755)
	at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:212)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at org.mortbay.jetty.bio.SocketConnector$Connection.run(SocketConnector.java:228)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
15:35:13.337 [                   refine] POST /command/core/importing-controller (4146ms)
15:35:15.675 [                   refine] POST /command/core/get-models (2338ms)
15:35:15.678 [                   refine] POST /command/core/get-rows (3ms)
```

### PR DESCRIPTION
**Pull request by wetneb**
The wikitable importer was making the assumption that all header cells were at the beginning of the table, which was unsafe. This removes this assumption: header cells are parsed like normal cells. The downside of this is that we do not detect automatically the headers: the user has to select how many lines of input should be parsed as header, just like in the CSV/TSV importer.

This also adds parsing for image links which are treated like templates (they can be either ignored or included in the table).

This should close #1448 - @ostephens let me know if that works for you.

### CODE REVIEW

**Comment by ostephens** [comment]
Still some problems with this I'm afraid.

The following wikitable works:
```
{| class="wikitable" style="width:80%; text-align:center;"
|-
! President
! Name
! Term of office start
! Term of office end
! Political Party
|-
| rowspan="2" align="center"|[[Carlos Menem]]
| '''[[Eduardo Bauzá]]'''
| 8 July 1995
| 5 June 1996
| [[Justicialist Party]]
|}
```
If I add in a colspan="2" in one of the column headers, with "Extract references in additional columns" checked, I get an error:
```
{| class="wikitable" style="width:80%; text-align:center;"
|-
! President
! Name
! colspan="2"|Term of office
! Political Party
|-
| rowspan="2" align="center"|[[Carlos Menem]]
| '''[[Eduardo Bauzá]]'''
| 8 July 1995
| 5 June 1996
| [[Justicialist Party]]
|}
```
and if I take the original table and add a header cell in one of the rows I also get an error:
```
{| class="wikitable" style="width:80%; text-align:center;"
|-
! President
! Name
! Term of office start
! Term of office end
! Political Party
|-
! rowspan="2" align="center"|[[Carlos Menem]]
| '''[[Eduardo Bauzá]]'''
| 8 July 1995
| 5 June 1996
| [[Justicialist Party]]
|}
```

In both cases if I uncheck the "Extract references in additional columns"  then the table parses successfully

**Comment by wetneb** [comment]
Arf - shameful! It shouldn't be hard to get right. Let me try again.

**Comment by wetneb** [comment]
@ostephens Your examples work fine here - are you sure you have properly recompiled OpenRefine when you tried the `issue1448` branch?

**Comment by ostephens** [comment]
Oh dear :( - let me check - apologies

**Comment by ostephens** [comment]
duh - forgot to clean before building - now successful

**Comment by wetneb** [comment]
Thanks a lot Owen!
